sklearn/utils/_typedefs.pxd

```diff
@@ -1,6 +1,30 @@
 #!python
 cimport numpy as cnp
 
+# Commonly used types
+# These are redefinitions of the ones defined by numpy in
+# https://github.com/numpy/numpy/blob/main/numpy/__init__.pxd
+# and exposed by cython in
+# https://github.com/cython/cython/blob/master/Cython/Includes/numpy/__init__.pxd.
+# It will eventually avoid having to always include the numpy headers even when we
+# would only use it for the types. (TODO don't cimport numpy in this extension)
+# 
+# When used to declare variables that will receive values from numpy arrays, it
+# should match the dtype of the array. For example, to declare a variable that will
+# receive values from a numpy array of dtype np.float64, the type float64_t must be
+# used.
+#
+# TODO: Stop defining custom types locally of globally like DTYPE_t and friends and
+# use these consistantly throughout the codebase.
+ctypedef unsigned char bool_t
+ctypedef Py_ssize_t intp_t
+ctypedef size_t uintp_t
+ctypedef signed int int32_t
+ctypedef signed long long int64_t
+ctypedef float float32_t
+ctypedef double float64_t
```

Comment:
I think we should also go ahead and include:
```python
ctypedef unsigned int       uint32_t
ctypedef unsigned long long uint64_t
```

since they appear in a few places in the code base.

Also, tiny nit in formatting, but what do you think about having the aliases aligned, e.g.
```python
ctypedef Py_ssize_t         intp_t
ctypedef size_t             uintp_t
ctypedef unsigned char      bool_t
ctypedef signed int         int32_t
ctypedef signed long long   int64_t
ctypedef float              float32_t
ctypedef double             float64_t
ctypedef unsigned int       uint32_t
ctypedef unsigned long long uint64_t
```

Reply:
> I think we should also go ahead and include

Actually I did the opposite and removed the ones I didn't need for ``_hierarchical_fast``. I'd rather not introduce dead code. I'd prefer that we define new types as needed when converting the other cython extensions

Reply:
> Also, tiny nit in formatting, but what do you think about having the aliases aligned, e.g.

This is not a formatting style we use in scikit-learn so I would not do it. But I'm not strongly opposed and if more devs prefer that I won't complain.